prom_sky_con/pkg/store_prom.go

```diff
@@ -0,0 +1,204 @@
+// For each flow reported, prepare prometheus entries, one for each direction of data flow.
+func (s *storePrometheus) StoreFlows(flows map[core.Tag][]interface{}) error {
+	secs := time.Now().Unix()
+	for _, val := range flows {
+		for _, i := range val {
+			f := i.(*flow.Flow)
+			if f.Transport == nil {
+				continue
+			}
+			initiator_ip := f.Network.A
+			target_ip := f.Network.B
+			initiator_port := strconv.FormatInt(f.Transport.A, 10)
+			target_port := strconv.FormatInt(f.Transport.B, 10)
+			node_tid := f.NodeTID
+			label1 := prometheus.Labels{
+				"initiator_ip":   initiator_ip,
+				"target_ip":      target_ip,
+				"initiator_port": initiator_port,
+				"target_port":    target_port,
+				"direction":      "initiator_to_target",
+				"node_tid":       node_tid,
+			}
+			label2 := prometheus.Labels{
+				"initiator_ip":   initiator_ip,
+				"target_ip":      target_ip,
+				"initiator_port": initiator_port,
+				"target_port":    target_port,
+				"direction":      "target_to_initiator",
+				"node_tid":       node_tid,
+			}
+			// post the info to prometheus
+			bytesSent.With(label1).Set(float64(f.Metric.ABBytes))
+			bytesSent.With(label2).Set(float64(f.Metric.BABytes))
+			cEntry := cacheEntry{
+				label1:    label1,
+				label2:    label2,
+				timeStamp: secs,
+			}
+			s.connectionCache.Set(f.L3TrackingID, cEntry, 0)
+		}
+	}
+	return nil
+}
+
+// cleanupExpiredEntries - any entry that has expired should be removed from the prometheus reporting and cache
+func (s *storePrometheus) cleanupExpiredEntries() {
+	secs := time.Now().Unix()
+	expireTime := secs - s.connectionTimeout
+	entriesMap := s.connectionCache.Items()
+	for k, v := range entriesMap {
+		v2 := v.Object.(cacheEntry)
+		if v2.timeStamp < expireTime {
+			// clean up the entry
+			logging.GetLogger().Debugf("secs = %s, deleting %s", secs, v2.label1)
+			bytesSent.Delete(v2.label1)
+			bytesSent.Delete(v2.label2)
+			s.connectionCache.Delete(k)
+		}
+	}
+}
+
+func (s *storePrometheus) cleanupExpiredEntriesLoop() {
+	for true {
+		s.cleanupExpiredEntries()
+		time.Sleep(cleanupTime * time.Second)
+	}
+}
+
+// registerCollector - needed in order to send metrics to prometheus
+func registerCollector(c prometheus.Collector) {
+	prometheus.Register(c)
+}
+
+// startPrometheusInterface listens for prometheus resource usage requests
+func startPrometheusInterface(s *storePrometheus) {
+
+	// Metrics have to be registered to be exposed:
+	registerCollector(bytesSent)
+
+	// The Handler function provides a default handler to expose metrics
+	// via an HTTP server. "/metrics" is the usual endpoint for that.
+	http.Handle("/metrics", promhttp.Handler())
+
+	err := http.ListenAndServe(s.port, nil)
+	if err != nil {
+		logging.GetLogger().Errorf("error on http.ListenAndServe = %s", err)
+		os.Exit(1)
+	}
+}
+
+// NewStorePrometheusInternal allocates and initializes the storePrometheus structure
+func NewStorePrometheusInternal(cfg *viper.Viper) (*storePrometheus, error) {
```

Comment:
rename to `newStorePrometheus`

Reply:
There is another method by that name which wraps this function. This internal function is separated from the wrapper so that it can be tested without making the threads that are created in the wrapper function.

Reply:
better to change such that `newStorePrometheus` is the internal function implementing `NewStorePrometheus` (this is an accepted convention in go).

Reply:
OK